vortex-expr/src/expr.rs

```diff
@@ -6,26 +6,23 @@ use std::sync::Arc;
 use vortex::array::{ConstantArray, StructArray};
 use vortex::compute::{compare, Operator as ArrayOperator};
 use vortex::variants::StructArrayTrait;
-use vortex::{Array, IntoArray};
+use vortex::{Array, ArrayDType, IntoArray};
 use vortex_dtype::field::Field;
-use vortex_dtype::DType;
-use vortex_error::{vortex_bail, vortex_err, VortexExpect as _, VortexResult, VortexUnwrap};
+use vortex_error::{vortex_err, VortexExpect as _, VortexResult};
 use vortex_scalar::Scalar;
-use vortex_schema::Schema;
 
 use crate::Operator;
 
-const NON_PRIMITIVE_COST_ESTIMATE: usize = 64;
-const COLUMN_COST_MULTIPLIER: usize = 1024;
-
 pub trait VortexExpr: Debug + Send + Sync + PartialEq<dyn Any> {
     fn as_any(&self) -> &dyn Any;
 
     fn evaluate(&self, batch: &Array) -> VortexResult<Array>;
 
     fn references(&self) -> HashSet<Field>;
 
-    fn estimate_cost(&self, schema: &Schema) -> usize;
+    fn project(&self, projection: &[Field]) -> Option<Arc<dyn VortexExpr>>;
+
+    fn is_constant(&self) -> bool;
```

Comment:
This should actually be `fn fold(&self) -> Option<Scalar>`, i.e. can this be evaluated to a constant?